numpy/core/src/multiarray/descriptor.c

```diff
@@ -1486,84 +1461,86 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
             }
             return NPY_FAIL;
         }
-        retval = PyArray_DescrConverter(obj2, at);
+        PyArray_Descr *ret = _convert_from_any(obj2);
         Py_DECREF(obj2);
-        return retval;
+        return ret;
     }
-
-    if (PyBytes_Check(obj)) {
-        return _convert_from_bytes(obj, at);
+    else if (PyBytes_Check(obj)) {
+        return _convert_from_bytes(obj);
     }
     else if (PyTuple_Check(obj)) {
         /* or a tuple */
-        *at = _convert_from_tuple(obj, 0);
-        if (*at == NULL){
-            if (!PyErr_Occurred()) {
-                _report_generic_error();
-            }
-            return NPY_FAIL;
-        }
-        return NPY_SUCCEED;
+        return _convert_from_tuple(obj, 0);
     }
     else if (PyList_Check(obj)) {
         /* or a list */
-        *at = _convert_from_array_descr(obj,0);
-        if (*at == NULL) {
-            if (!PyErr_Occurred()) {
-                _report_generic_error();
-            }
-            return NPY_FAIL;
-        }
-        return NPY_SUCCEED;
+        return _convert_from_array_descr(obj, 0);
     }
     else if (PyDict_Check(obj) || PyDictProxy_Check(obj)) {
         /* or a dictionary */
-        *at = _convert_from_dict(obj,0);
-        if (*at == NULL) {
-            if (!PyErr_Occurred()) {
-                _report_generic_error();
-            }
-            return NPY_FAIL;
-        }
-        return NPY_SUCCEED;
+        return _convert_from_dict(obj, 0);
     }
     else if (PyArray_Check(obj)) {
         _report_generic_error();
-        return NPY_FAIL;
+        return NULL;
     }
     else {
-        if (_arraydescr_from_dtype_attr(obj, at)) {
+        PyArray_Descr *ret;
+        if (_arraydescr_from_dtype_attr(obj, &ret)) {
             /*
              * Using dtype attribute, *at may be NULL if a
              * RecursionError occurred.
              */
-            if (*at == NULL) {
-                return NPY_FAIL;
-            }
-            return NPY_SUCCEED;
+            return ret;
         }
         /*
          * Note: this comes after _arraydescr_from_dtype_attr because the ctypes
          * type might override the dtype if numpy does not otherwise
          * support it.
          */
         if (npy_ctypes_check(Py_TYPE(obj))) {
-            *at = _arraydescr_from_ctypes_type(Py_TYPE(obj));
-            return *at ? NPY_SUCCEED : NPY_FAIL;
+            return _arraydescr_from_ctypes_type(Py_TYPE(obj));
         }
         _report_generic_error();
```

Comment:
```suggestion
```
Can you remove this here now?

Reply:
I could, but it does the opposite of what I'm trying to do - the goal is to have each of these helpers set their own error, rather than sloppily rely on the caller to check if we remembered to set an error.

This particular error is "I tried everything and nothing worked".

Reply:
True enough, may help to track down how to get rid of these cases.